src/main/java/duke/storage/Storage.java

```diff
@@ -0,0 +1,141 @@
+     *
+     * @return A duke.tasks.TaskList containing the tasks read from the storage file, or an empty
+     * duke.tasks.TaskList if the file is new.
+     */
+    private TaskList setupTasks() {
+        TaskList tasks = new TaskList();
+        try {
+            File directory = new File(directoryPath);
+            if (!directory.exists()) {
+                directory.mkdir();
+            }
+            File taskFile = new File(filePath);
+            if (taskFile.createNewFile()) {
+                //new file created
+            } else {
+                tasks = getTasks();
+            }
+
+        } catch (IOException e) {
+            System.out.println("Error occurred");
+            e.printStackTrace();
+        } finally {
+            return tasks;
+        }
+    }
+
+    /**
+     * Retrieves tasks from the storage file and returns them in a duke.tasks.TaskList.
+     *
+     * @return A duke.tasks.TaskList containing tasks read from the storage file.
+     */
+    public TaskList getTasks() {
+        TaskList taskList = new TaskList();
+        try {
+            Scanner scanner = new Scanner(new File(filePath));
+
+            while (scanner.hasNextLine()) {
+                String str = scanner.nextLine();
+                char taskType = str.charAt(1);
+                boolean isMarked = str.charAt(4) == 'X';
+                String description = str.split("] ")[1];
+                switch (taskType) {
+                case 'T':
+                    ToDo taskT = new ToDo(description);
+                    if (isMarked) {
+                        taskT.markAsDone();
+                    }
+                    taskList.addTask(taskT);
+                    break;
+                case 'D':
+                    description = description.split(" \\(by: ")[0];
+                    String by = str.split(" \\(by: ")[1];
+                    LocalDate date = LocalDate.parse(by.split(" ")[0]);
+                    LocalTime time = LocalTime.parse(by.split(" ")[1].replaceAll(".$", ""));
+                    Deadline taskD = new Deadline(description, date, time);
+                    if (isMarked) {
+                        taskD.markAsDone();
+                    }
+                    taskList.addTask(taskD);
+                    break;
+                case 'E':
+                    description = description.split(" \\(from: ")[0];
+                    String from = String.join("", str.split("\\(from: ")[1]).split(" to: ")[0];
+                    String to = str.split(" to: ")[1];
+                    LocalDate fromDate = LocalDate.parse(from.split(" ")[0]);
+                    LocalTime fromTime = LocalTime.parse(from.split(" ")[1]);
+                    LocalDate toDate = LocalDate.parse(to.split(" ")[0]);
+                    LocalTime toTime = LocalTime.parse(to.split(" ")[1].replaceAll(".$", ""));
+                    Event taskE = new Event(description, fromDate, fromTime, toDate, toTime);
```

Comment:
Maybe you could consider extracting out the functionalities within each case into its separate functions? So the switch block will not be so long as it could be hard to trace the different cases when they are so many lines apart?